lib/html5ever/precompiled.ex

```diff
@@ -0,0 +1,232 @@
+  @available_nif_versions ~w(2.14 2.15 2.16)
+
+  @doc """
+  Returns the target triple for download or compile and load.
+
+  This function is translating and adding more info to the system
+  architecture returned by Elixir/Erlang to one used by Rust.
+
+  The returning string format is the following:
+
+    "nif-NIF_VERSION-ARCHITECTURE-VENDOR-OS-ABI"
+
+  ## Examples
+
+      iex> target()
+      {:ok, "nif-2.16-x86_64-unknown-linux-gnu"} 
+
+      iex> target()
+      {:ok, "nif-2.15-aarch64-apple-darwin"}
+
+  """
+  def target(config \\ target_config()) do
+    sys_arch = String.split(config.system_arch, "-")
+
+    case config.os_type do
+      {:unix, osname} ->
+        arch_os = normalize_arch_os(osname, sys_arch) |> Enum.join("-")
+
+        cond do
+          arch_os not in @available_targets ->
+            {:error, "precompiled NIF is not available for this target: #{inspect(arch_os)}"}
+
+          config.nif_version not in @available_nif_versions ->
+            {:error,
+             "precompiled NIF is not available for this NIF version: #{inspect(config.nif_version)}"}
+
+          true ->
+            {:ok, "nif-#{config.nif_version}-#{arch_os}"}
+        end
+
+      {:win32, _} ->
+        {:error, "precompiled NIFs for Windows are not supported right now"}
+    end
+  end
+
+  defp target_config do
+    system_arch = :erlang.system_info(:system_architecture) |> List.to_string()
+
+    system_arch =
+      case @mix_target do
+        :host ->
+          system_arch
+
+        :x86_64 ->
+          system_arch
+
+        # Targets supported by the Nerves project.
+        # See: https://hexdocs.pm/nerves/targets.html
+        mix_target when mix_target in [:rpi, :rpi0, :rpi1, :rpi2, :rpi3, :rpi4, :bbb, :osd32mp1] ->
+          "arm-unknown-linux-gnueabihf"
+
+        other ->
+          raise "target #{inspect(other)} is not supported"
+      end
+
+    %{
+      os_type: :os.type(),
+      system_arch: system_arch,
+      nif_version: :erlang.system_info(:nif_version) |> List.to_string()
+    }
+  end
+
+  defp normalize_arch_os(:darwin, [arch, vendor | _rest]) do
+    arch = with "arm" <- arch, do: "aarch64"
+
+    [arch, vendor, "darwin"]
+  end
+
+  defp normalize_arch_os(:linux, [arch, vendor, os | maybe_abi]) do
+    arch = with "amd64" <- arch, do: "x86_64"
+    vendor = with "pc" <- vendor, do: "unknown"
+
+    # Fix vendor for Nerves
+    vendor =
+      if arch == "arm" and vendor == "buildroot" do
+        "unknown"
+      else
+        vendor
+      end
+
+    [arch, vendor, os | maybe_abi]
+  end
+
+  defp normalize_arch_os(_other, sys_arch), do: sys_arch
+
+  def download_or_reuse_nif_file(rustler_opts, opts) do
+    name = Keyword.fetch!(rustler_opts, :otp_app)
+    version = Keyword.fetch!(opts, :version)
+
+    priv_dir =
+      name
+      |> :code.priv_dir()
+      |> List.to_string()
+
+    cache_dir = :filename.basedir(:user_cache, Atom.to_string(name))
```

Comment:
For XLA/Torch, we are caching at `Path.join(File.cwd!(), "cache")`, instead of using a system cache.

  * Pros: we don't write outside of the project directory, this may also be easier to cache on CI?
  * Cons: each different project has to download it from scratch.

WDYT? @wojtekmach, any preferences?

Reply:
Not a strong opinion but so think the global cache would be nice to avoid duplication. We can use :filename.basedir with :user_cache argument. And if someone has `MIX_ENV` var set, we pass `os: :Linux` to use XDG. Wdyt?

Reply:
XLA stores it on the user_cache too, so this is good as is. In the future we can make the path customizable too, as in XLA: https://github.com/elixir-nx/xla/blob/main/lib/xla.ex#L59-L66